writer.go

```diff
@@ -180,3 +181,145 @@ func (w *FilteredLevelWriter) WriteLevel(level Level, p []byte) (int, error) {
+	// Destination writer. If LevelWriter is provided (usually), its WriteLevel is used
+	// instead of Write.
+	io.Writer
+
+	// ConditionalLevel is the level (and below) at which lines are buffered until
+	// a trigger level (or higher) line is emitted. Usually this is set to DebugLevel.
+	ConditionalLevel Level
+
+	// TriggerLevel is the lowest level that triggers the sending of the conditional
+	// level lines. Usually this is set to ErrorLevel.
+	TriggerLevel Level
+
+	buf       *bytes.Buffer
+	triggered bool
+	mu        sync.Mutex
+}
+
+func (w *TriggerLevelWriter) WriteLevel(l Level, p []byte) (n int, err error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.buf == nil {
+		return 0, errors.New("invalid writer")
+	}
+
+	// At first trigger level or above log line, we flush the buffer and change the
+	// trigger state to triggered.
+	if !w.triggered && l >= w.TriggerLevel {
+		err := w.trigger()
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// Unless triggered, we buffer everything at and below ConditionalLevel.
+	if !w.triggered && l <= w.ConditionalLevel {
+		// We prefix each log line with a byte with the level.
+		// Hopefully we will never have a level value which equals a newline
+		// (which could interfere with reconstruction of log lines in the trigger method).
+		w.buf.WriteByte(byte(l))
+		w.buf.Write(p)
+		return len(p), nil
+	}
+
+	// Anything above ConditionalLevel is always passed through.
+	// Once triggered, everything is passed through.
+	if lw, ok := w.Writer.(LevelWriter); ok {
+		return lw.WriteLevel(l, p)
+	}
+	return w.Write(p)
+}
+
+// trigger expects lock to be held.
+func (w *TriggerLevelWriter) trigger() error {
+	if w.triggered {
+		return nil
+	}
+	w.triggered = true
+	defer w.buf.Reset()
+
+	p := w.buf.Bytes()
+	for len(p) > 0 {
+		// We do not use bufio.Scanner here because we already have full buffer
+		// in the memory and we do not want extra copying from the buffer to
+		// scanner's token slice, nor we want to hit scanner's token size limit,
+		// and we also want to preserve newlines.
+		i := bytes.IndexByte(p, '\n')
+		line := p[0 : i+1]
+		p = p[i+1:]
+		// We prefixed each log line with a byte with the level.
+		level := Level(line[0])
+		line = line[1:]
+		var err error
+		if lw, ok := w.Writer.(LevelWriter); ok {
+			_, err = lw.WriteLevel(level, line)
+		} else {
+			_, err = w.Write(line)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Trigger forces flushing the buffer and change the trigger state to
+// triggered, if the writer has not already been triggered before.
+func (w *TriggerLevelWriter) Trigger() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.buf == nil {
+		return errors.New("invalid writer")
+	}
+
+	return w.trigger()
+}
+
+// Close closes the writer and returns the buffer to the pool.
+func (w *TriggerLevelWriter) Close() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.buf == nil {
+		return nil
+	}
+
+	w.buf.Reset()
+	triggerWriterPool.Put(w.buf)
```

Comment:
Should we add a size limit for returning the buffer to the pool? We may end up with some large buffer containing something like a stack trace we don't want to maintain in memory via the pool.

Reply:
I would guess that if one log line contains a stack trace, probably also other log lines from the same program contains them. So it is probably OK to have all buffers around as they are and grow to. Otherwise buffers will have to grow again and again to that size?

We could also make a setting in the writer field, for like max capacity to return to the pool with 0 (default) being that it returns everything.

Reply:
If you set that field, do we resize back to max capacity buffers before returning them to the pool (instead of dropping them completely)?

I would vote for keeping it as-is though.

Reply:
Imagine 1% of logs triggering a larger amount of logs, they will over time (pretty quickly actually) "pollute" the whole pool with large buffers and waste memory.

Those large buffers needs to be dropped, resizing won't change the memory taken by the backing array.

Reply:
So I add an option or we hard-code some value?

Reply:
OK, thank you for pointing this out. I read more about this now and it does not seem to be clear solution.

Related Golang issues:
* https://github.com/golang/go/issues/23199 (general issue that sync.Pool seems to be primarily suitable when stored data is of similar size)
* https://github.com/golang/go/issues/27735 (JSON encoding has a similar issue)

Some solutions seems to be:
* having a hard-coded limit as you proposed:
  * fmt standard library [has 64 KB limit](https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/fmt/print.go;l=169) and it seems it works well enough because most string formatting is on short strings
  * but trying to do a [similar limit for JSON encoding](https://go-review.googlesource.com/c/go/+/136035) had a large negative impact on performance and was abandoned and is [not yet solved](https://github.com/golang/go/issues/27735 (JSON) - I think our use case is closer to JSON encoding and less of fmt formatting
* do some local (per buffer) or global (per pool) statistics and decide which buffers to drop
  * https://github.com/valyala/bytebufferpool seems to do something like that, maybe there are also other 3rd party libraries to use here instead of trying to fix it ourselves
  * [this is an example of doing some local stats](https://github.com/golang/go/issues/27735#issuecomment-739169121) proposed for JSON encoding, but it seems it has not yet been found as the best approach
* do buffer segmentation (have different pools for different sizes of buffers) - not suitable for us because we do not know in advance the size we need

Reply:
You could implement `.Available()` with the API prior to Go 1.21. It's just cap minus len.

Reply:
I do not think you can, [looking at the code](https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/bytes/buffer.go;l=77-86) because you do not have `b.off`.

Reply:
Oh right. I have an intuition that grabbing buf.Length and buf.Capacity before buf.Reset and then grabbing them again after should give you enough information to figure out buf.Available but I'm not 100% sure rn.

Reply:
@rs So this is waiting for your input about the direction to take this.

Reply:
I would keep it simple and use a sensible limit.

Reply:
@rs: I have updated PR. It now drops the buffer if it grows too much. Default is 64 KB, based on fmt limit. But it can be changed if somebody wants/needs.